bittensor/utils/async_substrate.py

```diff
@@ -31,41 +60,340 @@ class Preprocessed:
+                    type_registry_name
+                )
+                self.type_registry_preset = type_registry_name
+            except ValueError:
+                type_registry_preset_dict = None
+
+        else:
+            type_registry_preset_dict = None
+
+        if type_registry_preset_dict:
+            # Load type registries in runtime configuration
+            if self.implements_scaleinfo is False:
+                # Only runtime with no embedded types in metadata need the default set of explicit defined types
+                self.runtime_config.update_type_registry(
+                    load_type_registry_preset(
+                        "legacy", use_remote_preset=use_remote_preset
+                    )
+                )
+
+            if self.type_registry_preset != "legacy":
+                self.runtime_config.update_type_registry(type_registry_preset_dict)
+
+        if self.type_registry:
+            # Load type registries in runtime configuration
+            self.runtime_config.update_type_registry(self.type_registry)
+
+
+class RequestManager:
+    RequestResults = dict[Union[str, int], list[Union[ScaleType, dict]]]
+
+    def __init__(self, payloads):
+        self.response_map = {}
+        self.responses = defaultdict(lambda: {"complete": False, "results": []})
+        self.payloads_count = len(payloads)
+
+    def add_request(self, item_id: int, request_id: Any):
+        self.response_map[item_id] = request_id
+
+    def overwrite_request(self, item_id: int, request_id: Any):
+        self.response_map[request_id] = self.response_map.pop(item_id)
+        return request_id
+
+    def add_response(self, item_id: int, response: dict, complete: bool):
+        request_id = self.response_map[item_id]
+        self.responses[request_id]["results"].append(response)
+        self.responses[request_id]["complete"] = complete
+
+    @property
+    def is_complete(self):
+        return (
+            all(info["complete"] for info in self.responses.values())
+            and len(self.responses) == self.payloads_count
+        )
+
+    def get_results(self):
+        return {
+            request_id: info["results"] for request_id, info in self.responses.items()
+        }
+
+
+class Websocket:
+    def __init__(
+        self,
+        ws_url: str,
+        max_subscriptions=1024,
+        max_connections=100,
+        shutdown_timer=5,
+        options: dict = None,
+    ):
+        """
+        Websocket manager object. Allows for the use of a single websocket connection by multiple
+        calls.
+
+        :param ws_url: Websocket URL to connect to
+        :param max_subscriptions: Maximum number of subscriptions per websocket connection
+        :param max_connections: Maximum number of connections total
+        :param shutdown_timer: Number of seconds to shut down websocket connection after last use
+        """
+        # TODO allow setting max concurrent connections and rpc subscriptions per connection
```

Comment:
Are these TODOs still relevant?

Reply:
Yes. I would like to see them added in the future, but they're not a necessary addition for the time being.